src/service.rs

```diff
@@ -331,45 +365,40 @@ where
             );
 
             loop {
-                let recv_attempt = self
-                    .global_state
-                    .read_arc()
-                    .await
-                    .response_receiver
-                    .try_recv();
-                if recv_attempt.is_ok() {
-                    break recv_attempt.map_err(|_| BuildError::Missing);
-                } else {
-                    let e = recv_attempt.unwrap_err();
-                    let is_empty = matches!(e, UnbounderTryRecvError::Empty);
-                    if is_empty {
+                match async_timeout(check_duration, response_receiver.recv()).await {
+                    Err(toe) => {
                         if Instant::now() >= timeout_after {
+                            tracing::warn!(%toe, "Couldn't get available blocks in time for parent {:?}",  req_msg.requested_vid_commitment);
                             // lookup into the builder_state_to_last_built_block, if it contains the result, return that otherwise return error
                             if let Some(last_built_block) = self
                                 .global_state
                                 .read_arc()
                                 .await
                                 .builder_state_to_last_built_block
-                                .get(for_parent)
+                                .get(&(*for_parent, view_num))
                             {
                                 tracing::info!(
                                     "Returning last built block for parent {:?}",
                                     req_msg.requested_vid_commitment
                                 );
                                 break Ok(last_built_block.clone());
                             }
-                            tracing::warn!(%e, "Couldn't get available blocks in time for parent {:?}",  req_msg.requested_vid_commitment);
                             break Err(BuildError::Error {
                                 message: "No blocks available".to_string(),
                             });
                         }
-                        async_compatibility_layer::art::async_yield_now().await;
                         continue;
-                    } else {
-                        tracing::error!(%e, "Channel closed while getting available blocks for parent {:?}", req_msg.requested_vid_commitment);
-                        break Err(BuildError::Error {
-                            message: "channel unexpectedly closed".to_string(),
-                        });
+                    }
+                    Ok(recv_attempt) => {
+                        if recv_attempt.is_ok() {
+                            break recv_attempt.map_err(|_| BuildError::Missing);
```

Comment:
IIUC this `map_err` is useless because we already know `recv_attempt.is_ok()`, and similarly the `unwrap_err()` below will never panic (but is an unnecessary red flag for readers). This would be much more clear written as a `match` on `recv_attempt`

Reply:
I'm responsible for the `.map_err`. It's strictly there to coerce the recv_attempt to the correct `Result<...>` type.

I could trivially remove it, but I didn't want to remove the tracing message.

Let me try an alternative....

Reply:
Is this acceptable?

Reply:
https://github.com/EspressoSystems/hotshot-builder-core/pull/120/commits/a395bb77c1ae86f458850830f090a74d29937e75